docs/development/systemtask.md

```diff
@@ -0,0 +1,263 @@
+Typical hazards and techniques of shared-memory multithreaded programming apply.
+
+A new task is most often needed when adding a fundamentally new subsystem
+to the Crazyflie firmware.
+
+If the new subsystem needs to do
+significant computation in response to inputs such as CRTP radio messages, it
+should receive those inputs on a queue and perform the computation within
+its own task instead of blocking the radio task.
+In this example, we will set up the skeleton for such a new subsystem.
+
+Development environment
+-----------------------
+
+You should have the
+[crazyflie-firmware](https://github.com/bitcraze/crazyflie-firmware) and
+[crazyflie-clients-python](https://github.com/bitcraze/crazyflie-clients-python)
+cloned in the same folder. If you are using the [Bitcraze
+VM](https://github.com/bitcraze/bitcraze-vm) this is already the case.
+
+The Crazyflie firmware should be on Master branch.
+
+For the rest of the howto you will work in the crazyflie-firmware
+project.
+
+
+Implementing the task
+---------------------
+
+High level tasks (those that do not directly talk to hardware) usually go in
+the `modules` directory. We will walk through the sections needed in a
+new file `modules/src/example.c`.
+
+
+First, include the necessary system header files:
+
+``` {.c}
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "static_mem.h"
+#include "task.h"
+```
+
+### File-static variables
+
+Next, we statically allocate the queue to hold inputs from
+other parts of the system such as the radio.
+Here we just hold integers in the queue. Often it will be a struct instead.
+A queue of length 1 is appropriate if new inputs supersede old ones.
+For example, if the input controls the current LED color,
+we do not care which other colors were requested in the past.
+
+Note that the `STATIC_MEM_*_ALLOC` family of macros expand to static
+variable declarations, and can be used outside of any function scope.
+
+``` {.c}
+static xQueueHandle inputQueue;
+STATIC_MEM_QUEUE_ALLOC(inputQueue, 1, sizeof(int));
+```
+
+
+Next, we allocate the task call stack and other necessary memory for OS
+bookkeeping. Note that `configMINIMAL_STACK_SIZE` gives a fairly small stack.
+Tasks that do lots of computation will almost certainly need larger stacks.
+
+The FreeRTOS task API requires that task functions take a `void *` argument.
+
+``` {.c}
+static void exampleTask(void*);
+STATIC_MEM_TASK_ALLOC(exampleTask, configMINIMAL_STACK_SIZE);
+```
+
+### Functions
+
+Next, we implement the `Init()` function.
+This should contain all of the FreeRTOS allocations and initializations
+that need to last as long as the firmware is running.
+Later in this howto, we will need to edit another file to call the `Init()`.
+
+``` {.c}
+static bool isInit = false;
+
+void exampleTaskInit() {
+  inputQueue = STATIC_MEM_QUEUE_CREATE(inputQueue);
+  // TODO
+  STATIC_MEM_TASK_CREATE(exampleTask, exampleTask, KALMAN_TASK_NAME, NULL, KALMAN_TASK_PRI);
```

Comment:
Copy/paste from the kalman task.
EXAMPLE_TASK_NAME and EXAMPLE_TASK_PRI would be better

Reply:
New constants created in 'config.h' in d3c0a97.